monitoring/alert/alert.go

```diff
@@ -0,0 +1,394 @@
+	// name.  channelNames keeps track of the new names.
+	channelNames := make(map[string]string)
+	for _, c := range b.Channels {
+		fmt.Fprintf(w, "Updating channel %q\n", c.GetDisplayName())
+		c.VerificationStatus = monitoringpb.NotificationChannel_VERIFICATION_STATUS_UNSPECIFIED
+		updated := false
+		if sameProject {
+			req := &monitoringpb.UpdateNotificationChannelRequest{
+				NotificationChannel: c.NotificationChannel,
+			}
+			_, err := channelClient.UpdateNotificationChannel(ctx, req)
+			if err == nil {
+				updated = true
+			}
+		}
+		if !updated {
+			req := &monitoringpb.CreateNotificationChannelRequest{
+				Name:                "projects/" + projectID,
+				NotificationChannel: c.NotificationChannel,
+			}
+			oldName := c.GetName()
+			c.Name = ""
+			newC, err := channelClient.CreateNotificationChannel(ctx, req)
+			if err != nil {
+				return err
+			}
+			channelNames[oldName] = newC.GetName()
+		}
+	}
+
+	for _, a := range b.Alerts {
+		fmt.Fprintf(w, "Updating alert %q\n", a.GetDisplayName())
+		a.CreationRecord = nil
+		a.MutationRecord = nil
+		for i, aChannel := range a.GetNotificationChannels() {
+			if c, ok := channelNames[aChannel]; ok {
+				a.NotificationChannels[i] = c
+			}
+		}
+		updated := false
+		if sameProject {
+			req := &monitoringpb.UpdateAlertPolicyRequest{
+				AlertPolicy: a.AlertPolicy,
+			}
+			_, err := alertClient.UpdateAlertPolicy(ctx, req)
+			if err == nil {
+				updated = true
+			}
+		}
+		if !updated {
+			req := &monitoringpb.CreateAlertPolicyRequest{
+				Name:        "projects/" + projectID,
+				AlertPolicy: a.AlertPolicy,
+			}
+			if _, err = alertClient.CreateAlertPolicy(ctx, req); err != nil {
+				log.Fatal(err)
+			}
+		}
+	}
+	fmt.Fprintf(w, "Successfully restored alerts.")
+	return nil
+}
+
+// [END monitoring_alert_restore_policies]
+// [END monitoring_alert_create_policy]
+// [END monitoring_alert_create_channel]
+// [END monitoring_alert_update_channel]
+
+// [START monitoring_alert_replace_channels]
+
+// replaceChannels replaces the notification channels in the alert policy
+// with channelIDs.
+func replaceChannels(w io.Writer, projectID, alertPolicyID string, channelIDs []string) error {
+	ctx := context.Background()
+
+	client, err := monitoring.NewAlertPolicyClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	a := &monitoringpb.AlertPolicy{
+		Name: "projects/" + projectID + "/alertPolicies/" + alertPolicyID,
+	}
+	for _, c := range channelIDs {
+		c = "projects/" + projectID + "/notificationChannels/" + c
+		a.NotificationChannels = append(a.NotificationChannels, c)
+	}
+	req := &monitoringpb.UpdateAlertPolicyRequest{
+		AlertPolicy: a,
+		UpdateMask: &fieldmask.FieldMask{
+			Paths: []string{"notification_channels"},
+		},
+	}
+	if _, err := client.UpdateAlertPolicy(ctx, req); err != nil {
+		return fmt.Errorf("UpdateAlertPolicy: %v", err)
+	}
+	fmt.Fprintf(w, "Successfully replaced channels.")
+	return nil
+}
+
+// [END monitoring_alert_replace_channels]
```

Comment:
is this region tag intentionally outside the function block? seems inconsistent with the earlier samples

Reply:
Yes. Reused sample code is inside the function for a couple, but most are outside (which seems consistent with most other samples.)